conformance/tests/backendtlspolicy.go

```diff
@@ -0,0 +1,65 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tests
+
+import (
+	"testing"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	"sigs.k8s.io/gateway-api/conformance/utils/http"
+	"sigs.k8s.io/gateway-api/conformance/utils/kubernetes"
+	"sigs.k8s.io/gateway-api/conformance/utils/suite"
+	"sigs.k8s.io/gateway-api/pkg/features"
+)
+
+func init() {
+	ConformanceTests = append(ConformanceTests, BackendTLSPolicy)
+}
+
+var BackendTLSPolicy = suite.ConformanceTest{
+	ShortName:   "BackendTLSPolicy",
+	Description: "A single service that is targeted by a BackendTLSPolicy must successfully complete TLS termination",
+	Features: []features.SupportedFeature{
+		features.SupportGateway,
+		features.SupportBackendTLSPolicy,
+	},
+	Manifests: []string{"tests/backendtlspolicy.yaml"},
+	Test: func(t *testing.T, suite *suite.ConformanceTestSuite) {
+		ns := "gateway-conformance-infra"
+		routeNN := types.NamespacedName{Name: "gateway-conformance-infra-test", Namespace: ns}
+		gwNN := types.NamespacedName{Name: "gateway-backendtlspolicy", Namespace: ns}
+
+		kubernetes.NamespacesMustBeReady(t, suite.Client, suite.TimeoutConfig, []string{ns})
+
+		gwAddr := kubernetes.GatewayAndHTTPRoutesMustBeAccepted(t, suite.Client, suite.TimeoutConfig, suite.ControllerName, kubernetes.NewGatewayRef(gwNN), routeNN)
+
+		// Verify that the response to a call to /backendTLS will return the matching SNI.
+		t.Run("Simple request targeting BackendTLSPolicy should reach infra-backend", func(t *testing.T) {
+			http.MakeRequestAndExpectEventuallyConsistentResponse(t, suite.RoundTripper, suite.TimeoutConfig, gwAddr,
+				http.ExpectedResponse{
+					Request: http.Request{
+						Path: "/backendTLS",
+					},
+					Response:  http.Response{StatusCode: 200},
+					Backend:   "infra-backend-v1",
+					Namespace: "gateway-conformance-infra",
+					SNI:       "abc.example.com",
+				})
+		})
```

Comment:
Digging into the `MakeRequestAndExpectEventuallyConsistentResponse` code one thing I noticed is that it appears that is always sends the `http://` scheme?

Meaning: we're sending requests to the Gateway in the clear, and then proving that the Gateway added the correct backend TLS configuration on the backend and then served the response back out in the clear.

So (if I'm not misunderstanding anything) that's a perfectly valid test but to me it kind of begs the question: just for _completeness_ should we not also test when the frontend connection has a distinct TLS configuration and **verify the fingerprint on the clientside, and then verify the different fingerprint on the server side** just to to thoroughly prove that the implementations are working properly?

Note that when I dug into this I was in a bit of a rush because I had a meeting about to start, so it's possible I'm misunderstanding, but take a look and LMKWYT and I'll come back and take another look at this myself soon :thinking:

Reply:
Yes, I agree, except we have a separate test for frontend TLS in [TLSRouteSimpleSameNamespace](https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/tests/tlsroute-simple-same-namespace.go#L41)  This PR adds a test just for backend TLS.  Adding a test for both frontend and backend TLS is needed.

There are a couple of subtests that are missing.  I'd like to get just one merged before following up with the others.  This is my to do list:
- add a test case where the backend tls validation is misconfigured and ensure it returns an error
- add a test case where frontend connection is also TLS
- add a test case where frontend connection is also TLS and misconfigured

Reply:
Ok cool, thanks for the heads up. Those test cases sound good, if you think it's feasible I might just add the point that we should test the fingerprints at each level as this helps to ensure more definitively everything is working exactly as expected.

I'm very much down with getting one valuable test in for now and follow up with the others. I would just ask if they're not already, get those test cases laid out in a follow-up issue somewhere, and then please feel free to consider this comment resolved at your discretion. :+1:

Reply:
I'm not sure if you saw that the test cases were laid out, debated, and updated in #3138.  And based on your feedback I can add:

- [ ] verify fingerprints on both frontend connection and backend connection, where each has a distinct but correct TLS configuration.  (Use edge-terminated TLS and BackendTLSPolicy together)
- [ ] misconfigure frontend tls validation and ensure it returns an error
- [ ] misconfigure backend tls validation and ensure it returns an error

Do I understand your point?

Reply:
Yes. The current tests are valid, but there's certainly more cases we can cover.

Reply:
I think those additional test cases make sense. And if the intention is to add those iteratively as follow-up PRs, I am OK with that as long as they're tracked in #3138, and if so please feel free to consider this comment resolved at your discretion.

Reply:
In the last commit https://github.com/kubernetes-sigs/gateway-api/pull/3212/commits/7b08c849badbcc6cf753c39a246c0c88bcd043a9, I changed this to make an https request instead of http.